features/lorawan/lorastack/phy/LoRaPHY.cpp

```diff
@@ -122,316 +100,1315 @@ void LoRaPHY::handle_send(uint8_t *buf, uint8_t size)
+{
+    bool dr_invalid = false;
+    bool freq_invalid = false;
+
+    if (!phy_params.custom_channelplans_supported
+            || id >= phy_params.max_channel_cnt) {
+
+        return LORAWAN_STATUS_PARAMETER_INVALID;
+    }
+
+    // Validate the datarate range
+    if (val_in_range(new_channel->dr_range.fields.min,
+                     phy_params.min_tx_datarate,
+                     phy_params.max_tx_datarate) == 0) {
+        dr_invalid = true;
+    }
+
+    if (val_in_range(new_channel->dr_range.fields.max, phy_params.min_tx_datarate,
+                     phy_params.max_tx_datarate) == 0) {
+        dr_invalid = true;
+    }
+
+    if (new_channel->dr_range.fields.min > new_channel->dr_range.fields.max) {
+        dr_invalid = true;
+    }
+
+    // Default channels don't accept all values
+    if (id < phy_params.default_channel_cnt) {
+        // Validate the datarate range for min: must be DR_0
+        if (new_channel->dr_range.fields.min > phy_params.min_tx_datarate) {
+            dr_invalid = true;
+        }
+
+        // Validate the datarate range for max: must be DR_5 <= Max <= TX_MAX_DATARATE
+        if (val_in_range(new_channel->dr_range.fields.max,
+                         phy_params.default_max_datarate,
+                         phy_params.max_tx_datarate) == 0) {
+            dr_invalid = true;
+        }
+
+        // We are not allowed to change the frequency
+        if (new_channel->frequency != phy_params.channels.channel_list[id].frequency) {
+            freq_invalid = true;
+        }
+    }
+
+    // Check frequency
+    if (!freq_invalid) {
+        if (verify_frequency(new_channel->frequency) == false) {
+            freq_invalid = true;
+        }
+    }
+
+    // Check status
+    if (dr_invalid && freq_invalid) {
+        return LORAWAN_STATUS_FREQ_AND_DR_INVALID;
+    }
+
+    if (dr_invalid) {
+        return LORAWAN_STATUS_DATARATE_INVALID;
+    }
+
+    if (freq_invalid) {
+        return LORAWAN_STATUS_FREQUENCY_INVALID;
+    }
+
+    memcpy(&(phy_params.channels.channel_list[id]), new_channel, sizeof(channel_params_t));
+
+    phy_params.channels.channel_list[id].band = new_channel->band;
+
+    // if there are multiple channel masks, i.e., there are more than 16 channels
+    // defined by the PHY layer, we search the channel index in all masks and
+    // set the appropriate bit in the appropriate mask
+    for (uint8_t i = 0; i < phy_params.max_channel_cnt; i ++) {
+        if (i == id) {
+            MASK_BIT_SET(phy_params.channels.mask[i/16], (i % 16));
+        }
+    }
+
+    return LORAWAN_STATUS_OK;
+}
+
+bool LoRaPHY::remove_channel(uint8_t channel_id)
+{
+    // upper layers are checking if the custom channel planning is supported or
+    // not. So we don't need to worry about that
+    for (uint8_t i = 0; i < phy_params.channels.mask_size; i++) {
+        if (MASK_BIT_TEST(phy_params.channels.default_mask[i], channel_id) != 0) {
```

Comment:
Ah, found it. This was the comment that got lost yesterday evening. This is broken. Code errors are being introduced because of the difficulty of manipulating the bit array, which is why I'm still banging on about the helper macro. Once I see no more bugs, I may drop it...